clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1006,7 +1004,14 @@ public void assign(Collection<TopicPartition> partitions) {
         fetcher.sendFetches();
 
         long now = time.milliseconds();
-        client.poll(Math.min(coordinator.timeToNextPoll(now), timeout), now);
+        long pollTimeout = Math.min(coordinator.timeToNextPoll(now), timeout);
+
+        client.poll(pollTimeout, now, new PollCondition() {
+            @Override
+            public boolean pollNeeded() {
+                return !fetcher.hasCompletedFetches();
+            }
+        });
```

Comment:
One more question, what if the fetch returns from the background thread, but it failed instead of succeeding? In that case, we won't have any new completed fetches and we will block until the timeout. Is that what we want?

Also, I am not sure if you saw my question about the heart-beating that is mentioned in the method documentation.